### ISSUE TITLE
Add diagnostic setting to prevent `str` matching against `Iterable[str]`

### ISSUE THREAD

**Issue by kylebebak** [microsoft/pyright#4886]
## Problem

This enhancement request comes from an old, much discussed [issue in python/typing](https://github.com/python/typing/issues/256). Consider a function like this one:

```py
def send_emails(emails: Iterable[str]):
    for email in emails:
        send_email(email)
```

Client code should call this with something like `send_emails(["a@example.com", "b@example.com"])`, or `send_emails(*get_emails())`, but the following will also pass type checks: `send_emails("a@example.com")`.

The last way of calling this function will lead to bad behavior at runtime, and the type system can't catch it, because `str` is also `Iterable[str]`. We could modify the signature of `send_emails` so it only takes `List[str]`, but that's both overly restrictive and overly permissive. It prevents other natural ways of calling the function, and allows the function to do things to `emails` that it probably shouldn't, like append a value to it.

## Solution

[Do what pytype does](https://github.com/google/pytype/blob/main/docs/faq.md#why-doesnt-str-match-against-string-iterables).

> As of early August 2021, Pytype introduced a check that forbids matching str against the following types to prevent a common accidental bug:
> 
> Iterable[str], Sequence[str], Collection[str], Container[str]

pyright could add a diagnostic setting to enable this same type checking behavior.

If a user wants to write an alternate `send_emails` implementation that also accepts `str`, they can still annotate it like so:

```py
def send_emails(emails: str | Iterable[str]):
    if isinstance(emails, str):
        emails = [emails]
    for email in emails:
        send_email(email)
```

## Context

There's a lot of good context in the aforementioned issue in python/typing, e.g. [here](https://github.com/python/typing/issues/256#issuecomment-440830800). `pytype` added this config option as of mid 2021.

**Comment by erictraut** [comment]
I've been following the thread in python/typing, and I don't think there's a consensus here. I was hoping that a consensus would emerge, and the solution would be implemented in typeshed so it is common to all Python type checkers.

I'm reluctant to add a diagnostic for this because it's not generally a type error and is not necessarily indicative of a bug to use a `str` in this manner. It's unfortunate that the language allowed a `str` to be used directly as a sequence rather than forcing users to call a method on the `str` to retrieve a sequence.

I'm also reluctant to override the behavior indicated by the typeshed stubs with special-cased logic for this type. Any time we've done that in the past, I've regretted it because it inevitably leads to robustness and maintenance issues in pyright.

I'm going to close this request for now, but I'm willing to reconsider it in the future if there is sufficient interest from pyright users (in the form of upvotes) or the python/typing discussion converges on a consensus.
